nwb_conversion_tools/utils/spike_interface.py

```diff
@@ -1148,160 +1155,37 @@ def write_units(
-                elif isinstance(prop_value, dict):
-                    print(f"Skipping property '{pr}' because dictionaries are not supported.")
-                    skip_properties.append(pr)
-                    break
-            else:
-                shapes.append(np.nan)
-        property_shapes[pr] = shapes
-
-    for pr in property_shapes.keys():
-        elems = [elem for elem in property_shapes[pr] if not np.any(np.isnan(elem))]
-        if not np.all([elem == elems[0] for elem in elems]):
-            print(f"Skipping property '{pr}' because it has variable size across units.")
-            skip_properties.append(pr)
-
+    # add properties
+    all_properties = checked_sorting.get_property_keys()
     write_properties = set(all_properties) - set(skip_properties)
     for pr in write_properties:
         if pr not in property_descriptions:
             warnings.warn(
                 f"Description for property {pr} not found in property_descriptions. "
                 "Setting description to 'no description'"
             )
+    aggregated_unit_properties = defaultdict()
     for pr in write_properties:
         unit_col_args = dict(name=pr, description=property_descriptions.get(pr, "No description."))
         if pr in ["max_channel", "max_electrode"] and nwbfile.electrodes is not None:
             unit_col_args.update(table=nwbfile.electrodes)
         units_table.add_column(**unit_col_args)
-
-    aggregated_unit_properties = defaultdict(list)
-    for unit_id in unit_ids:
-
-        for pr in write_properties:
-            if pr in sorting.get_unit_property_names(unit_id):
-                aggregated_unit_properties[pr].append(sorting.get_unit_property(unit_id, pr))
-            else:  # Case of missing data for this unit and this property
-                aggregated_unit_properties[pr].append(None)
-
-    # handle missing data differently depending on type of data
-    for key, val in aggregated_unit_properties.items():
-        if all(isinstance(x, int) or x is None for x in val) and any(x is None for x in val):
-            aggregated_unit_properties[key] = [np.nan if x is None else float(x) for x in val]
-        if all(isinstance(x, str) or x is None for x in val):
-            aggregated_unit_properties[key] = [x or "" for x in val]
-        if all(isinstance(x, float) or x is None for x in val):
-            aggregated_unit_properties[key] = [np.nan if x is None else x for x in val]
+        aggregated_unit_properties[pr] = checked_sorting.get_property(pr)
 
     for i, unit_id in enumerate(unit_ids):
         if use_times:
-            spkt = sorting.frame_to_time(sorting.get_unit_spike_train(unit_id=unit_id))
-        else:
-            spkt = sorting.get_unit_spike_train(unit_id=unit_id) / sorting.get_sampling_frequency()
+            warnings.warn(
+                "'use_times' is not supported yet by the new SI API. " "Computing spike times using sampling frequency"
```

Comment:
@alejoe91 Would it make sense to delay this refactor until the new SI API supports everything it used to? What new features or improvements do the new Sorting objects bring?

Reply:
I believe only times are missing for the sorting object. Will work on it next week and we can wait until then 🙂

Reply:
@CodyCBakerPhD this implements the time mechanism: https://github.com/SpikeInterface/spikeinterface/pull/464